File: internal/appupdate/appupdate.go
// Package appupdate keeps the Windows app current against the Tokitoki
// server. Check asks whether a newer build exists; Install downloads it next
// to the running executable, verifies the published digest, and swaps the
// binary in place. Windows never lets a running executable be overwritten,
// but it does let it be renamed — so the old binary steps aside as ".old"
// and the new one takes its path, exactly as the CLI's self-update does.
package appupdate

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
	"time"
)

// Channel is the release channel the Windows app reads on the Tokitoki server.
const Channel = "windows"

const (
	checkTimeout    = 15 * time.Second
	downloadTimeout = 10 * time.Minute

	// oldSuffix marks the previous binary, which the running process cannot
	// delete but can rename aside. The leftover is removed on the next app
	// start or the next install, whichever comes first.
	oldSuffix = ".old"
)

// ErrDevBuild marks a local build, which has no meaningful version to compare.
var ErrDevBuild = errors.New("development build; update checks are disabled")

// semverRE accepts what the server's update check accepts.
var semverRE = regexp.MustCompile(`^v?[0-9]+\.[0-9]+\.[0-9]+`)

// Update describes an available newer build.
type Update struct {
	// Version is the offered release version.
	Version string
	// DownloadURL is the absolute URL the build can be fetched from.
	DownloadURL string
	// Size is the asset's byte count; 0 when the server did not publish one.
	Size int64
	// SHA256 is the asset's hex digest. Present whenever the release
	// published one; the download is rejected if it does not match.
	SHA256 string
}

// Check returns the available update, or nil when current is already the
// newest published build.
func Check(ctx context.Context, baseURL, current string) (*Update, error) {
	if !semverRE.MatchString(current) {
		return nil, ErrDevBuild
	}

	ctx, cancel := context.WithTimeout(ctx, checkTimeout)
	defer cancel()

	checkURL := fmt.Sprintf(
		"%s/api/updates/check?channel=%s&platform=windows&arch=%s&version=%s",
		baseURL, Channel, runtime.GOARCH, url.QueryEscape(current),
	)
	response, err := httpGet(ctx, checkURL)
	if err != nil {
		return nil, fmt.Errorf("check for update: %w", err)
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("check for update: server returned %s", response.Status)
	}

	var decoded struct {
		Available bool   `json:"available"`
		Version   string `json:"version"`
		URL       string `json:"url"`
		Size      int64  `json:"size"`
		SHA256    string `json:"sha256"`
	}
	if err := json.NewDecoder(response.Body).Decode(&decoded); err != nil {
		return nil, fmt.Errorf("check for update: %w", err)
	}
	if !decoded.Available {
		return nil, nil
	}
	if decoded.URL == "" {
		return nil, fmt.Errorf("check for update: update %s has no download URL", decoded.Version)
	}
	return &Update{
		Version:     decoded.Version,
		DownloadURL: baseURL + decoded.URL,
		Size:        decoded.Size,
		SHA256:      decoded.SHA256,
	}, nil
}

// Install downloads the update and replaces the running executable. On
// return the path holds the new build: the next launch runs it, whether the
// user restarts now or not. The running process keeps executing the old
// image either way.
func Install(ctx context.Context, update *Update) error {
	executable, err := os.Executable()
	if err != nil {
		return fmt.Errorf("locate executable: %w", err)
	}
	return install(ctx, update, executable)
}

// install is Install with the executable made explicit, so tests can point
// it at a scratch file.
func install(ctx context.Context, update *Update, executable string) error {
	// The downloaded bytes become the next thing this machine executes.
	// Fetching them over cleartext HTTP hands that to whoever sits on the
	// path — loopback is the only exception, because it never leaves the
	// machine.
	if err := requireTrustedTransport(update.DownloadURL); err != nil {
		return err
	}

	// A leftover from a previous swap; gone unless that process still runs.
	_ = os.Remove(executable + oldSuffix)

	ctx, cancel := context.WithTimeout(ctx, downloadTimeout)
	defer cancel()

	tmp, err := download(ctx, update, filepath.Dir(executable))
	if err != nil {
		return err
	}
	defer os.Remove(tmp)

	return swap(tmp, executable)
}

// download streams the new binary into a temp file inside dir — the
// executable's own directory, so the final rename never crosses volumes.
func download(ctx context.Context, update *Update, dir string) (string, error) {
	response, err := httpGet(ctx, update.DownloadURL)
	if err != nil {
		return "", fmt.Errorf("download update: %w", err)
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return "", fmt.Errorf("download update: server returned %s", response.Status)
	}

	file, err := os.CreateTemp(dir, ".tokitoki-update-*")
	if err != nil {
		return "", fmt.Errorf("download update: %w", err)
	}
	digest := sha256.New()
	written, err := io.Copy(io.MultiWriter(file, digest), response.Body)
	if closeErr := file.Close(); err == nil {
		err = closeErr
	}
	if err == nil && update.Size > 0 && written != update.Size {
		err = fmt.Errorf("got %d bytes, expected %d", written, update.Size)
	}
	// The digest ties the bytes on disk to the bytes the release was
	// published with, before anything executes them. Releases without a
	// published digest fall back to the size check.
	if err == nil && update.SHA256 != "" {
		got := hex.EncodeToString(digest.Sum(nil))
		want := strings.ToLower(strings.TrimPrefix(update.SHA256, "sha256:"))
		if got != want {
			err = fmt.Errorf("sha256 mismatch: downloaded %s, server published %s", got, want)
		}
	}
	if err == nil {
		err = os.Chmod(file.Name(), 0o755)
	}
	if err != nil {
		_ = os.Remove(file.Name())
		return "", fmt.Errorf("download update: %w", err)
	}
	return file.Name(), nil
}

// swap installs the new binary over the old one: the old binary is renamed
// aside, the new one renamed onto its path. Both renames are atomic, and a
// failed second rename puts the old binary back — the executable path never
// ends up empty or torn.
func swap(newBinary, executable string) error {
	old := executable + oldSuffix
	if err := os.Rename(executable, old); err != nil {
		return fmt.Errorf("install update: %w", err)
	}
	if err := os.Rename(newBinary, executable); err != nil {
		_ = os.Rename(old, executable)
		return fmt.Errorf("install update: %w", err)
	}
	return nil
}

// requireTrustedTransport rejects download URLs an on-path attacker can
// rewrite.
func requireTrustedTransport(rawURL string) error {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return fmt.Errorf("update download URL %q: %w", rawURL, err)
	}
	if parsed.Scheme == "https" {
		return nil
	}
	host := parsed.Hostname()
	if host == "localhost" {
		return nil
	}
	if ip := net.ParseIP(host); ip != nil && ip.IsLoopback() {
		return nil
	}
	return fmt.Errorf("refusing to download an update over %q: a non-loopback update server must use https", rawURL)
}

// CleanupLeftovers removes the ".old" binary a previous update left beside
// the executable. Failure is normal — the previous process may still be
// exiting — and the next install retries anyway.
func CleanupLeftovers() {
	if executable, err := os.Executable(); err == nil {
		_ = os.Remove(executable + oldSuffix)
	}
}

// Relaunch starts a fresh process from executable. The caller must release
// the single-instance lock first, or the new process sees a live instance
// and quits.
func Relaunch(executable string) error {
	return exec.Command(executable).Start()
}

func httpGet(ctx context.Context, target string) (*http.Response, error) {
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, target, nil)
	if err != nil {
		return nil, err
	}
	request.Header.Set("User-Agent", "tokitoki-windows")
	return http.DefaultClient.Do(request)
}

File: src/app_update.rs
//! App self-update: check, download, verify, and atomic binary swap
//! (mirrors Go `internal/appupdate`). There is no external updater process —
//! the running exe renames itself aside and puts the new one in its place.

use std::{
    fs, io,
    io::{Read, Write},
    net::IpAddr,
    path::{Path, PathBuf},
    time::Duration,
};

use serde::Deserialize;
use sha2::{Digest, Sha256};

/// Update channel reported to the server.
pub const CHANNEL: &str = "windows";
const CHECK_TIMEOUT: Duration = Duration::from_secs(15);
const DOWNLOAD_TIMEOUT: Duration = Duration::from_secs(10 * 60);
const OLD_SUFFIX: &str = ".old";
const USER_AGENT: &str = "tokitoki-windows";

/// Failures across the update pipeline.
#[derive(Debug, thiserror::Error)]
pub enum Error {
    /// The running binary is a dev build; updates are disabled.
    #[error("development build; updates are disabled")]
    DevBuild,
    /// The update-check request failed or returned an unusable payload.
    #[error("update check failed: {0}")]
    Check(String),
    /// The download URL uses a transport we refuse (non-https, non-loopback).
    #[error("untrusted update transport: {0}")]
    UntrustedTransport(String),
    /// Downloading the new binary failed.
    #[error("update download failed: {0}")]
    Download(String),
    /// The downloaded size differs from what the server announced.
    #[error("update size mismatch: expected {expected} bytes, got {got}")]
    SizeMismatch {
        /// Bytes the server announced.
        expected: u64,
        /// Bytes actually received.
        got: u64,
    },
    /// The downloaded bytes don't hash to the announced SHA-256.
    #[error("update digest mismatch")]
    DigestMismatch,
    /// Filesystem work (temp file, rename swap) failed.
    #[error("update install failed: {0}")]
    Install(#[from] io::Error),
}

/// An available update as announced by the server.
#[derive(Clone, Debug)]
pub struct Update {
    /// Version string, e.g. `0.2.0`.
    pub version: String,
    /// Absolute download URL (server returns a relative path; we prefix base).
    pub download_url: String,
    /// Expected size in bytes; `0` disables the size check.
    pub size: u64,
    /// Expected SHA-256 hex digest (optionally `sha256:`-prefixed); empty
    /// disables the digest check.
    pub sha256: String,
}

#[derive(Deserialize)]
struct CheckResponse {
    #[serde(default)]
    available: bool,
    #[serde(default)]
    version: String,
    #[serde(default)]
    url: String,
    #[serde(default)]
    size: u64,
    #[serde(default)]
    sha256: String,
}

/// Asks the server whether an update is available for `current`.
///
/// # Errors
/// Returns [`Error::DevBuild`] when `current` is not `x.y.z`-shaped, or
/// [`Error::Check`] when the request fails.
pub fn check(base_url: &str, current: &str) -> Result<Option<Update>, Error> {
    if !is_semver_ish(current) {
        return Err(Error::DevBuild);
    }
    let mut url = url::Url::parse(&format!("{base_url}/api/updates/check"))
        .map_err(|err| Error::Check(err.to_string()))?;
    url.query_pairs_mut()
        .append_pair("channel", CHANNEL)
        .append_pair("platform", "windows")
        .append_pair("arch", arch())
        .append_pair("version", current);

    let response = ureq::AgentBuilder::new()
        .timeout(CHECK_TIMEOUT)
        .user_agent(USER_AGENT)
        .build()
        .get(url.as_str())
        .call()
        .map_err(|err| Error::Check(err.to_string()))?;
    let decoded: CheckResponse = response
        .into_json()
        .map_err(|err| Error::Check(err.to_string()))?;

    if !decoded.available {
        return Ok(None);
    }
    Ok(Some(Update {
        version: decoded.version,
        download_url: format!("{base_url}{}", decoded.url),
        size: decoded.size,
        sha256: decoded.sha256,
    }))
}

/// Downloads and installs `update` next to the running exe, swapping the
/// current binary aside as `<exe>.old`. The app keeps running; the new binary
/// takes effect on relaunch.
///
/// # Errors
/// Returns [`Error`] on untrusted transport, download/verification failure,
/// or a failed rename swap (which is rolled back).
pub fn install(update: &Update) -> Result<(), Error> {
    let exe = std::env::current_exe()?;
    require_trusted_transport(&update.download_url)?;

    // A leftover .old from a previous update would block the swap.
    let old = old_path(&exe);
    let _ = fs::remove_file(&old);

    let staged = download(update, &exe)?;
    swap(&exe, &old, &staged)
}

/// Deletes a leftover `<exe>.old` from a previous update. Best-effort.
pub fn cleanup_leftovers() {
    if let Ok(exe) = std::env::current_exe() {
        let _ = fs::remove_file(old_path(&exe));
    }
}

/// Starts `target` as a detached process; the caller must have released the
/// single-instance mutex first or the child will see a live instance and quit.
///
/// # Errors
/// Returns an [`io::Error`] when the process cannot be spawned.
pub fn relaunch(target: &Path) -> io::Result<()> {
    std::process::Command::new(target).spawn().map(drop)
}

fn old_path(exe: &Path) -> PathBuf {
    let mut name = exe.as_os_str().to_owned();
    name.push(OLD_SUFFIX);
    PathBuf::from(name)
}

/// `x86_64` / `aarch64` → the Go-style arch tokens the server expects.
fn arch() -> &'static str {
    match std::env::consts::ARCH {
        "x86_64" => "amd64",
        "aarch64" => "arm64",
        other => other,
    }
}

fn is_semver_ish(version: &str) -> bool {
    crate::agent_cli::parse_version(version.split(['-', '+']).next().unwrap_or(version)).is_some()
}

/// https is always trusted; plain http only for localhost/loopback (dev).
fn require_trusted_transport(raw_url: &str) -> Result<(), Error> {
    let untrusted = || Error::UntrustedTransport(raw_url.to_owned());
    let parsed = url::Url::parse(raw_url).map_err(|_| untrusted())?;
    match parsed.scheme() {
        "https" => Ok(()),
        "http" => {
            let host = parsed.host_str().ok_or_else(untrusted)?;
            let loopback = host.eq_ignore_ascii_case("localhost")
                || host.parse::<IpAddr>().is_ok_and(|ip| ip.is_loopback());
            loopback.then_some(()).ok_or_else(untrusted)
        }
        _ => Err(untrusted()),
    }
}

fn download(update: &Update, exe: &Path) -> Result<PathBuf, Error> {
    let dir = exe.parent().unwrap_or(Path::new("."));
    let staged = dir.join(format!(".tokitoki-update-{}", std::process::id()));

    let response = ureq::AgentBuilder::new()
        .timeout(DOWNLOAD_TIMEOUT)
        .user_agent(USER_AGENT)
        .build()
        .get(&update.download_url)
        .call()
        .map_err(|err| Error::Download(err.to_string()))?;

    let result = stream_verified(response.into_reader(), &staged, update);
    if result.is_err() {
        let _ = fs::remove_file(&staged);
    }
    result.map(|()| staged)
}

fn stream_verified(mut body: impl Read, staged: &Path, update: &Update) -> Result<(), Error> {
    let mut file = fs::File::create(staged)?;
    let mut hasher = Sha256::new();
    let mut written: u64 = 0;
    let mut buffer = vec![0u8; 64 * 1024];
    loop {
        let read = body
            .read(&mut buffer)
            .map_err(|err| Error::Download(err.to_string()))?;
        if read == 0 {
            break;
        }
        file.write_all(&buffer[..read])?;
        hasher.update(&buffer[..read]);
        written += read as u64;
    }
    file.sync_all()?;
    drop(file);

    if update.size > 0 && written != update.size {
        return Err(Error::SizeMismatch {
            expected: update.size,
            got: written,
        });
    }
    let expected = normalize_digest(&update.sha256);
    if !expected.is_empty() {
        let got = format!("{:x}", hasher.finalize());
        if got != expected {
            return Err(Error::DigestMismatch);
        }
    }
    Ok(())
}

/// Strips an optional `sha256:` prefix and lowercases the hex digest.
fn normalize_digest(digest: &str) -> String {
    let trimmed = digest.trim();
    let stripped = trimmed
        .get(..7)
        .filter(|prefix| prefix.eq_ignore_ascii_case("sha256:"))
        .map_or(trimmed, |_| &trimmed[7..]);
    stripped.to_ascii_lowercase()
}

/// Renames `exe` aside and the staged download into place; rolls the original
/// back if the second rename fails. Windows can't overwrite a running exe but
/// can rename it.
fn swap(exe: &Path, old: &Path, staged: &Path) -> Result<(), Error> {
    fs::rename(exe, old)?;
    if let Err(err) = fs::rename(staged, exe) {
        let _ = fs::rename(old, exe);
        let _ = fs::remove_file(staged);
        return Err(err.into());
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    mod is_semver_ish {
        use super::*;

        #[test]
        fn accepts_release_versions() {
            assert!(is_semver_ish("0.1.0"));
        }

        #[test]
        fn accepts_v_prefix() {
            assert!(is_semver_ish("v1.2.3"));
        }

        #[test]
        fn rejects_dev() {
            assert!(!is_semver_ish("dev"));
        }
    }

    mod require_trusted_transport {
        use super::*;

        #[test]
        fn accepts_https() {
            assert!(require_trusted_transport("https://tokitoki.dev/x").is_ok());
        }

        #[test]
        fn accepts_http_localhost() {
            assert!(require_trusted_transport("http://localhost:3000/x").is_ok());
        }

        #[test]
        fn accepts_http_loopback_ip() {
            assert!(require_trusted_transport("http://127.0.0.1/x").is_ok());
        }

        #[test]
        fn rejects_plain_http_hosts() {
            assert!(require_trusted_transport("http://evil.example/x").is_err());
        }
    }

    mod normalize_digest {
        use super::*;

        #[test]
        fn strips_sha256_prefix_case_insensitively() {
            assert_eq!(normalize_digest("SHA256:ABCDEF"), "abcdef");
        }

        #[test]
        fn lowercases_bare_digests() {
            assert_eq!(normalize_digest("ABCDEF"), "abcdef");
        }
    }

    mod swap {
        use super::*;

        #[test]
        fn replaces_exe_and_keeps_backup() {
            let dir = tempfile::tempdir().unwrap();
            let exe = dir.path().join("app.exe");
            let old = dir.path().join("app.exe.old");
            let staged = dir.path().join("staged");
            fs::write(&exe, b"v1").unwrap();
            fs::write(&staged, b"v2").unwrap();

            swap(&exe, &old, &staged).unwrap();

            assert_eq!(fs::read(&exe).unwrap(), b"v2");
        }

        #[test]
        fn rolls_back_when_staged_binary_is_missing() {
            let dir = tempfile::tempdir().unwrap();
            let exe = dir.path().join("app.exe");
            let old = dir.path().join("app.exe.old");
            fs::write(&exe, b"v1").unwrap();

            let result = swap(&exe, &old, &dir.path().join("missing"));

            assert!(result.is_err());
            assert_eq!(fs::read(&exe).unwrap(), b"v1");
        }
    }
}
